protocol/tx/entry.go

```diff
@@ -11,20 +11,44 @@ import (
 	"chain/protocol/bc"
 )
 
-type entry interface {
-	Type() string
-	Body() interface{}
-
-	// When an entry is created from a bc.TxInput or a bc.TxOutput, this
-	// reports the position of that antecedent object within its
-	// transaction. Both inputs (spends and issuances) and outputs
-	// (including retirements) are numbered beginning at zero. Entries
-	// not originating in this way report -1.
-	Ordinal() int
-}
+type (
+	entry interface {
+		Type() string
+		Body() interface{}
+
+		// When an entry is created from a bc.TxInput or a bc.TxOutput, this
+		// reports the position of that antecedent object within its
+		// transaction. Both inputs (spends and issuances) and outputs
+		// (including retirements) are numbered beginning at zero. Entries
+		// not originating in this way report -1.
+		Ordinal() int
+	}
+
+	// EntryRef holds an entry and its ID. The entry may be nil, but the
+	// ID of a valid entry must still be present.
+	EntryRef struct {
+		entry
+		bc.Hash
+	}
+)
 
 var errInvalidValue = errors.New("invalid value")
 
+// NewEntryRef produces a new EntryRef with a non-nil entry.
+func NewEntryRef(e entry) *EntryRef {
```

Comment:
It will be. I had intended that the next PR after this one would export a bunch of stuff from this package (in preparation for one following that would move everything into `protocol/bc`, in further preparation for using these types in `protocol/vm`).

I could save the exporting for the last minute before it's needed, or I can anticipate the need, which will keep the later PRs smaller. I don't have too strong a preference about that, LMK if you do.